### ISSUE TITLE
Rawtherapee 5.8 is not able to view CR2 files generated with CHDK.

### ISSUE THREAD

**Issue by crass** [Beep6581/RawTherapee#6490]
**Short description**

I have some CR2 files generated with CHDK that I'm trying to view in RawTherapee. I had used another program to convert them to DNG, so I have the CR2 files and the DNG in the same directory. When I view the directory in Raw Therapee only the DNG files are shown. When I run RawTherapee with a CR2 file Raw Therapee opens to image editor/viewer but its blank (not white or black, but the RawTherapee theme background color).

**Steps to reproduce**
1. RawTherapee-5.8.appimage <CR2 file>

**Expected behavior**

I expect to see the CR2 images with preview in the directory listing / File Browser tab. And when giving the file directly to Raw Therapee at start, I expect to see the image displayed.

**Additional information**
- Version of RawTherapee: Appimage 5.8
- Version of your operating system: Ubuntu 22.04
- `ParseExtensions=jpg;tif;tiff;png;crw;cr2;crf;nef;raf;pef;dng;arw;sr2;mrw;raw;orf;kdc;rwz;rw2;mef;`
- In case I've got a misconfiguration elsewhere: [options.gz](https://github.com/Beep6581/RawTherapee/files/8749885/options.gz)
- Test image: [IMG_1169.CR2.gz](https://github.com/Beep6581/RawTherapee/files/8749868/IMG_1169.CR2.gz)

**Comment by Lawrence37** [comment]
I wonder if CHDK is creating CR2 files in a special format. Several programs besides RawTherapee are unable to open it. Even ExifTool and Exiv2 report problems trying to read the metadata.

**Comment by crass** [comment]
Yes, [CHDK CR2](https://chdk.fandom.com/wiki/Raw#General_information) is not the CR2 created natively by Canon cameras. There are other programs that do view the attached file fine, eg. [nomacs](https://nomacs.org/) latest version. I believe this is because it uses libraw, which as far as I can tell RawTherapee does not. Does RawTherapee have its own code to raw file decoding, or is it using another library?

What's interesting is that [libraw does not support this camera](https://www.libraw.org/supported-cameras-libraw-202110), yet its still displaying.

**Comment by Lawrence37** [comment]
RawTherapee uses a tweaked version of dcraw. LibRaw is also based on dcraw. If LibRaw supports CHDK CR2s, it may be straightforward to port the relevant parts to RawTherapee. But based on what you wrote, it sounds like LibRaw just happens to work fine? I'm not familiar with nomas. Is it actually decoding the raw file or displaying the embedded JPEG preview?

**Comment by crass** [comment]
[Libraw lists some supported camera's](https://www.libraw.org/supported-cameras-libraw-202110) that use the CHDK hack, so I'm guessing it does support CHDK CR2s. I'm not super familiar with nomacs either. I have set a setting in Settings -> Advanced -> "RAW Loader settings" that says "Always Load Raw Data" as opposed to "Always Load JPG if Embedded". I'm not sure if despite that, its using the embedded JPEG. I'm having a hard time extracting the JPEG to see.

**Comment by arturoisilvia** [comment]
I have a Canon IXUS 310 HS with CHDK and the dng's it generates have been working fine in RT for some time now.

**Comment by crass** [comment]
Yes, DNG is a standard format, I wouldn't expect RT to have problems parsing DNGs (though it might not fully parse them). The issue in question is the CR2s produced by CHDK. @arturoisilvia, can you verify that a CR2 produced by your camera running CHDK is not recognized by RT?

**Comment by arturoisilvia** [comment]
My camera is a low-end camera. Without chdk it only does jpg and with chdk it can do dng+jpg. But not cr2.

Edit: I just played with my ixus and I can do cr2 with chdk. I have been able to open it with RT without problems.
However I'm sticking with dng because cr2 doesn't record exif data for me.

**Comment by Entropy512** [comment]
> If LibRaw supports CHDK CR2s, it may be straightforward to port the relevant parts to RawTherapee.

Longterm it might be beneficial to migrate to libraw like ART did - the legacy dcraw codebase is a bit...   meh...

That would definitely be a post-5.9 thing though as it's a massive rework.

**Comment by crass** [comment]
> However I'm sticking with dng because cr2 doesn't record exif data for me.

Yes, I think sticking with DNG makes the most sense as the CR2 is just a raw sensor dump, which should be embedded in the DNG anyway. My issue is that I already have CHDK generated CR2 raw files.

> I just played with my ixus and I can do cr2 with chdk. I have been able to open it with RT without problems.

Thanks for verifying that RT does read your CHDK generated CR2s. I wonder why it won't read mine.
